src/expr.rs

```diff
@@ -348,32 +630,111 @@ pub enum BinaryOperation<'a> {
+    /// in the last chain.
+    ///
+    /// If issued in a base chain, it is equivalent to the base chain policy.
     Return,
+    /// Continue evaluation at the first rule in chain.
+    ///
+    /// The current position in the ruleset is pushed to a call stack and
+    /// evaluation will continue there when the new chain is entirely evaluated
+    /// or a [return](Verdict::Return) verdict is issued. In case an absolute
+    /// verdict is issued by a rule in the chain, ruleset evaluation terminates
+    /// immediately and the specific action is taken.
     Jump(JumpTarget<'a>),
+    /// Similar to jump, but the current position is not pushed to the call
+    /// stack, meaning that after the new chain evaluation will continue at the
+    /// last chain instead of the one containing the goto statement.
     Goto(JumpTarget<'a>),
 }
 
 #[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
 #[serde(rename = "elem")]
 /// Explicitly set element object.
+///
+/// Element-related commands allow one to change contents of named
+/// [sets](crate::schema::Set) and [maps](crate::schema::Map).
 pub struct Elem<'a> {
+    /// The element value.
     pub val: Box<Expression<'a>>,
+    /// Timeout value for [sets](crate::schema::Set)/[maps](crate::schema::Map).
+    /// with flag [timeout](crate::schema::SetFlag::Timeout)
     pub timeout: Option<u32>,
+    /// The time until given element expires, useful for ruleset replication only.
     pub expires: Option<u32>,
+    /// Per element comment field.
     pub comment: Option<Cow<'a, str>>,
+    /// Enable a [counter][crate::stmt::Counter] per element.
+    ///
+    /// Added in nftables version *0.9.5*.
     pub counter: Option<Counter<'a>>,
 }
 
+/// Default impl for [Elem].
+impl Default for Elem<'_> {
+    fn default() -> Self {
+        Elem {
+            val: Box::new(Expression::String("10.2.3.4".into())),
+            timeout: None,
+            expires: None,
+            comment: None,
+            counter: None,
+        }
+    }
+}
+
 #[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
 #[serde(rename = "socket")]
 /// Construct a reference to packet’s socket.
 pub struct Socket<'a> {
-    pub key: Cow<'a, str>,
+    /// The socket attribute to match on.
+    pub key: Cow<'a, SocketAttr>,
+}
+
+/// Default impl for [Socket] with [wildcard](SocketAttr::Wildcard) key.
+impl Default for Socket<'_> {
+    fn default() -> Self {
+        Socket {
+            key: Cow::Borrowed(&SocketAttr::Wildcard),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash, Serialize, Deserialize)]
+#[serde(rename_all = "lowercase")]
+/// A [socket][Socket] attribute to match on.
+pub enum SocketAttr {
+    /// Match on the `IP_TRANSPARENT` socket option in the found socket.
+    Transparent,
+    /// Match on the socket mark (`SOL_SOCKET`, `SO_MARK`).
+    Mark,
+    /// Indicates whether the socket is wildcard-bound (e.g. 0.0.0.0 or ::0).
```

Comment:
I think `::` does not need a 0 suffix, right?

```suggestion
    /// Indicates whether the socket is wildcard-bound (e.g. 0.0.0.0 or ::).
```

Reply:
You are right, the compressed form is `::0`.
I do prefer the explicit form though for the documentation purpose.